inmem/kvlock.go

```diff
@@ -0,0 +1,127 @@
+package inmem
+
+import "context"
+
+//  This is an RWMutex using channels that allows for context timeouts.
+//  It is adapted from Rob Pike's blog post:
+//  https://blogtitle.github.io/go-advanced-concurrency-patterns-part-3-channels/
+// where he made an RWMutex using channels.  I just added the context handlng.
+// h/t to Bryan Mills for pointing out this blog post to me.
+type RWMutex struct {
+	write   chan struct{}
+	readers chan int
+}
+
+func NewLock() RWMutex {
+	return RWMutex{
+		// This is used as a normal Mutex.
+		write: make(chan struct{}, 1),
+		// This is used to protect the readers count.
+		// By receiving the value it is guaranteed that no
+		// other goroutine is changing it at the same time.
+		readers: make(chan int, 1),
+	}
+}
+
+func (l RWMutex) Lock() { l.write <- struct{}{} }
+
+func (l RWMutex) TryLock(ctx context.Context) bool {
+	if ctx == nil {
+		select {
+		case l.write <- struct{}{}:
+			return true
+		default:
+			return false
+
+		}
+	}
+	select {
+	case l.write <- struct{}{}:
+		return true
+	case <-ctx.Done():
+		return false
+	}
+}
+
+func (l RWMutex) Unlock() { <-l.write }
+
+func (l RWMutex) RLock() {
+	// Count current readers. Default to 0.
+	var rs int
+	// Select on the channels without default.
+	// One and only one case will be selected and this
+	// will block until one case becomes available.
+	select {
+	case l.write <- struct{}{}: // One sending case for write.
+		// If the write lock is available we have no readers.
+		// We grab the write lock to prevent concurrent
+		// read-writes.
+	case rs = <-l.readers: // One receiving case for read.
+		// There already are readers, let's grab and update the
+		// readers count.
+	}
+	// If we grabbed a write lock this is 0.
+	rs++
+	// Updated the readers count. If there are none this
+	// just adds an item to the empty readers channel.
+	l.readers <- rs
+}
+
+func (l RWMutex) RTryLock(ctx context.Context) bool {
+	// Count current readers. Default to 0.
+	var rs int
+	// Select on the channels without default.
+	// One and only one case will be selected and this
+	// will block until one case becomes available.
+	if ctx == nil {
```

Comment:
could you simplify logic and:

```
if ctx == nil {
 ctx = context.Background()
}
```
?

So that you write the logci only once?

Reply:
I thought about it, but it seemed wasteful to wait on a channel that would never return.